### PR TITLE
Adding HttpMethod and Header support

### PR DESCRIPTION
**Pull request by mathewc**
Adding the following:

* Allow Http methods to specify their supported Http methods in metadata via a new "methods" array property
* By default, all methods are allowed as before
* FunctionController honors the declared methods, and returns 405 for disallowed methods
* Adding Node.js support for reading request headers and writing response headers
* Adding general script support for reading request headers (see the BAT file example)

Addresses https://github.com/Azure/azure-webjobs-sdk-script/issues/13

### CODE REVIEW

File: sample/HttpTrigger/function.json [hunk 1]
```diff
@@ -2,7 +2,8 @@
   "bindings": [
     {
       "type": "httpTrigger",
-      "direction": "in"
+      "direction": "in",
+      "methods": [ "get" ]
```
Comment (davidebbo) [inline sample/HttpTrigger/function.json]: Maybe leave one sample without explicit methods to make sure that still works.

File: sample/HttpTrigger/index.js [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
+        context.log('test-header=' + headerValue);
+    }
+
     if (typeof req.query.name == 'undefined') {
         context.done(null, "Please pass a name on the query string");
     }
     else {
         context.done(null, "Hello " + req.query.name);
-    }
+    }
```
Comment (davidebbo) [inline sample/HttpTrigger/index.js]: yank spaces

File: sample/HttpTrigger/function.json (hunk 1 above)
Comment (mathewc) [inline sample/HttpTrigger/function.json]: Sure. I'll remove the specification from the BAT sample.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs [hunk 3]
```diff
@@ -42,7 +42,7 @@ public FunctionsController(WebScriptHostManager scriptHostManager, WebHookReceiv
 
             // Authorize the request
             SecretManager secretManager = (SecretManager)controllerContext.Configuration.DependencyResolver.GetService(typeof(SecretManager));
-            HttpBindingMetadata httpFunctionMetadata = (HttpBindingMetadata)function.Metadata.InputBindings.FirstOrDefault(p => p.Type == BindingType.HttpTrigger);
+            HttpTriggerBindingMetadata httpFunctionMetadata = (HttpTriggerBindingMetadata)function.Metadata.InputBindings.FirstOrDefault(p => p.Type == BindingType.HttpTrigger);
```
Comment (davidebbo) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Isn't http always intrinsically a trigger? What's the different between the two?

File: sample/HttpTrigger/function.json (hunk 1 above)
Comment (mathewc) [inline sample/HttpTrigger/function.json]: The BAT example in the e2e tests doesn't have a methods specification. I've tested it both ways.

File: sample/HttpTrigger/index.js (hunk 2 above)
Comment (mathewc) [inline sample/HttpTrigger/index.js]: I think this is just a github display issue. Looking at the file in VS there are no extraneous lines/spaces

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 3 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: No, there is the Http output binding - they're distinct.

File: sample/HttpTrigger/index.js (hunk 2 above)
Comment (mathewc) [inline sample/HttpTrigger/index.js]: Done.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 3 above)
Comment (davidebbo) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: I see. I still get confused about this. Maybe a side question, but is it fair to say that you can never have an http output binding without an http trigger input binding?

File: src/WebJobs.Script/Binding/HttpBinding.cs [hunk 4]
```diff
@@ -55,10 +55,19 @@ public override CustomAttributeBuilder GetCustomAttribute()
...
 
                     response = new HttpResponseMessage(statusCode);
                     response.Content = new StringContent(body);
+
+                    JObject headers = (JObject)jsonObject["headers"];
+                    if (headers != null)
+                    {
+                        foreach (var header in headers)
+                        {
+                            response.Headers.Add(header.Key, header.Value.ToString());
```
Comment (davidebbo) [inline src/WebJobs.Script/Binding/HttpBinding.cs]: Could `Value` be null and crash here?

File: src/WebJobs.Script/Description/NodeFunctionInvoker.cs [hunk 5]
```diff
@@ -297,6 +297,15 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
             inputDictionary["method"] = request.Method.ToString().ToUpperInvariant();
             inputDictionary["query"] = request.GetQueryNameValuePairs().ToDictionary(kv => kv.Key, kv => kv.Value, StringComparer.OrdinalIgnoreCase);
 
+            Dictionary<string, string> headers = new Dictionary<string, string>();
```
Comment (davidebbo) [inline src/WebJobs.Script/Description/NodeFunctionInvoker.cs]: nit: `var` when type is fully explicit on right side.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 3 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Currently, yes. The http out binding for http triggered functions should really be optional in the metadata file - we should just default one if not present.

File: src/WebJobs.Script/Binding/HttpBinding.cs (hunk 4 above)
Comment (mathewc) [inline src/WebJobs.Script/Binding/HttpBinding.cs]: I'll check and add a guard if so.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 3 above)
Comment (davidebbo) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: Maybe quick issue to track this as future improvement? It's lower pri at this point.

File: src/WebJobs.Script/Description/NodeFunctionInvoker.cs (hunk 5 above)
Comment (mathewc) [inline src/WebJobs.Script/Description/NodeFunctionInvoker.cs]: Ah, I use var sometimes, but I think this is a personal preference thing. We have pretty strict stylecop in place in this repo and it likes my code, so I'll leave as is.

File: src/WebJobs.Script/Description/ScriptFunctionInvoker.cs [hunk 6]
```diff
@@ -180,6 +180,12 @@ private void InitializeEnvironmentVariables(Dictionary<string, string> environme
                     string varName = string.Format(CultureInfo.InvariantCulture, "REQ_QUERY_{0}", queryParam.Key.ToUpperInvariant());
                     environmentVariables[varName] = queryParam.Value;
                 }
+
+                foreach (var header in request.Headers)
+                {
+                    string varName = string.Format(CultureInfo.InvariantCulture, "REQ_HEADERS_{0}", header.Key.ToUpperInvariant());
```
Comment (davidebbo) [inline src/WebJobs.Script/Description/ScriptFunctionInvoker.cs]: Is the purpose of those env variables to flow things to batch-link languages?

Comment (mathewc) [inline src/WebJobs.Script/Description/ScriptFunctionInvoker.cs]: Yep, see e2e tests below. We're following the same pattern we do now for query params.

File: src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs (hunk 3 above)
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs]: https://github.com/Azure/azure-webjobs-sdk-script/issues/60

File: src/WebJobs.Script/Binding/HttpBinding.cs (hunk 4 above)
Comment (mathewc) [inline src/WebJobs.Script/Binding/HttpBinding.cs]: Added a guard

File: src/WebJobs.Script/HttpMethodJsonConverter.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+                return true;
+            }
+            return false;
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException("reader");
```
Comment (davidebbo) [inline src/WebJobs.Script/HttpMethodJsonConverter.cs]: Isn't it a contract that Newtonsoft.Json will pass this? Kind of a nit, but not a fan of adding unneeded parameter checks in too many places, as they do add up in IL size (and bloat sources).

Comment (mathewc) [inline src/WebJobs.Script/HttpMethodJsonConverter.cs]: It addresses an fxcop error. We can discuss outside of this review - if we want to remove that fxcop check then we won't get such errors anymore.

**Comment by davidebbo** [comment]
Looks good other than minor comments. Consider making some changes to https://github.com/Azure/azure-webjobs-sdk-templates for a few templates that call for it. We should also discuss whether specifying methods is UI worthy.

File: src/WebJobs.Script/HttpMethodJsonConverter.cs (hunk 7 above)
Comment (davidebbo) [inline src/WebJobs.Script/HttpMethodJsonConverter.cs]: :cop: :hankey:

**Comment by davidebbo** [comment]
I assume the methods changes is forward compatible, such that if you get it in templates before runtime, it won't blow up?

File: src/WebJobs.Script/HttpMethodJsonConverter.cs (hunk 7 above)
Comment (mathewc) [inline src/WebJobs.Script/HttpMethodJsonConverter.cs]: :cop: :+1:

**Comment by mathewc** [comment]
Yes, no explosions either way. If the template has methods, but the runtime isn't looking for it all is good. If the runtime supports methods but templates don't specify, all is ok.

**Pull request closed without merging**